src/Chapter3.hs

```diff
@@ -1006,6 +1136,39 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data BraveKnight = BraveKnight
+  { kAttack :: Int
+  , kHealth :: Int
+  , kDefence :: Int
+  } deriving (Show)
+
+data FuriousMonster = FuriousMonster
+  { mHealth :: Int
+  , mAttack :: Int
+  } deriving (Show)
+
+-- unused datatypes
+data KnightActions = KnightAttack | DrinkPotion | CastSpell
+data MonsterActions = MonsterAttack | Runaway
+
+class Fighter a b where
+  residualHealthKnight :: a -> b -> a
+  residualHealthMonster :: a -> b -> b
+
+instance Fighter BraveKnight FuriousMonster where
```

Comment:
Unfortunately, with this design you need to implement all four combinations of typeclasses
```suggestion
instance Fighter BraveKnight FuriousMonster where
instance Fighter FuriousMonster BraveKnight where
instance Fighter BraveKnight BraveKnight where
instance Fighter FuriousMonster FuriousMonster where
```

to be able to implement a final polymorphic function:

```haskell
duel ::  (Fighter a b) => a -> b -> Result
```

Unfortunately, you can notice, that having 4 instances results into a lot of boilerplate code :disappointed: